src/Microsoft.AspNetCore.Server.IISIntegration/Server/IO/WebSocketsAsyncIOEngine.cs

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Server.IISIntegration
+{
+    internal partial class WebSocketsAsyncIOEngine: IAsyncIOEngine
+    {
+        private readonly IntPtr _handler;
+
+        private bool _isInitialized;
+
+        private AsyncInitializeOperation _initializationFlush;
+
+        private WebSocketWriteOperation _cachedWebSocketWriteOperation;
+
+        private WebSocketReadOperation _cachedWebSocketReadOperation;
+
+        private AsyncInitializeOperation _cachedAsyncInitializeOperation;
+
+        public WebSocketsAsyncIOEngine(IntPtr handler)
+        {
+            _handler = handler;
+        }
+
+        public ValueTask<int> ReadAsync(Memory<byte> memory)
+        {
+            CheckInitialized();
+
+            var read = GetReadOperation();
+            read.Initialize(_handler, memory);
+            read.Invoke();
+            return new ValueTask<int>(read, 0);
+        }
+
+        public ValueTask<int> WriteAsync(ReadOnlySequence<byte> data)
+        {
+            var write = GetWriteOperation();
+            write.Initialize(_handler, data);
+            write.Invoke();
+            return new ValueTask<int>(write, 0);
+        }
+
+        public ValueTask FlushAsync()
+        {
+            if (_isInitialized)
+            {
+                return new ValueTask(Task.CompletedTask);
+            }
+
+            NativeMethods.HttpEnableWebsockets(_handler);
+
+            var init = GetInitializeOperation();
+            init.Initialize(_handler);
+
+            var continuation = init.Invoke();
+
+            if (continuation != null)
+            {
+                _isInitialized = true;
+            }
+            else
+            {
+                _initializationFlush = init;
+            }
+
+            return new ValueTask(init, 0);
+        }
+
+        public void NotifyCompletion(int hr, int bytes)
+        {
+            _isInitialized = true;
+
+            var init = _initializationFlush;
+            if (init == null)
+            {
+                throw new InvalidOperationException("Unexpected completion for WebSocket operation");
+            }
+
+            var continuation = init.Complete(hr, bytes);
+
+            _initializationFlush = null;
+
+            continuation.Invoke();
+        }
+
+        private void CheckInitialized()
+        {
```

Comment:
CheckInitialize is really checking that flush was called before Read/Write, correct? I'd rename it to "ThrowIfNotInitialized" or "ThrowIfNotFlushed"

Reply:
+1